trie/utils/db.py

```diff
@@ -0,0 +1,63 @@
+import contextlib
+
+MISSING = object()
+
+
+class ScratchDB:
+    """
+    A wrapper of basic DB objects with uncommitted DB changes stored in local cache,
+    which represents as a dictionary of database keys and values.
+    This class should be usable as a context manager, the changes either all fail or all succeed.
+    Upon exiting the context, it writes all of the key value pairs from the cache into
+    the underlying database. If anyerror occurred before committing phase,
+    we would not apply commits at all.
+    """
+    def __init__(self, wrapped_db):
+        self.wrapped_db = wrapped_db
+        self.cache = {}  # type: Dict[bytes, bytes]
+        self.deletes = []
+
+    #
+    # Dictionary API
+    #
+    # if not key is found, return None
+    def __getitem__(self, key):
+        if key in self.cache:
+            return self.cache[key]
+        else:
+            try:
+                current_value = self.wrapped_db.get(key)
+            except KeyError:
+                current_value = None
+            return current_value
+
+    def __setitem__(self, key, value):
+        self.cache[key] = value
+
+    def __delitem__(self, key):
+        val = self.cache.pop(key, MISSING)
+        if val is MISSING:
```

Comment:
I don't understand this.  `val == Missing` when the key is not present in the cache.  Why do we only add it to the deletes set if the key was missing?  Seems like it should be added on any delete?  I'm missing something.

Reply:
Either the value was in the cache or the wrapped_db. If the value was added to the cache, then we delete it by removing it from the cache. Otherwise, we queue it up for deletion from the wrapped database at commit time.

Reply:
```python
wrapped_db = {b'a': 4}
cached_db = CachedDB(wrapped_db)
cached_db[b'a'] = 3  # b'a' is now in the cache, db[b'a'] -> 3
del cached_db[b'a']  # b'a' is not in the cache, b'a' not in delete set, db[b'a'] -> 4  (wrong!)
```

What am I missing?

Reply:
Because the db of a trie always stores `db[keccak(val)] = val`, having different key/values in the wrapped and cached db would mean finding two different pre-images with the same sha3 hash. I don't know if there's only one right thing to do in that situation, but leaving the old one there when pruning the new one doesn't seem insane (the only cost being a little extra database usage). I also don't see anything incorrect about deleting always, but it causes a performance hit:

1. We carry around a lot more deletes to try to run on the wrapped db
2. We would now have to test if the key exists before every delete operation on `wrapped_db`

I could go either way, and I don't think it's hard to change later. Thoughts?